### DISCUSSION TITLE
Question: intent of schema license definition required properties

### DISCUSSION THREAD

**Discussion by jhoward-lm** [CycloneDX/specification#493] [Q&A] [answered]
I have a clarification question about the intent of the required properties in the `license` definitions in the JSON schemas ([link](https://github.com/CycloneDX/specification/blob/62a669075f1897193a14060e0784e6a7576b693d/schema/bom-1.6.schema.json#L1229)). The same question applies to other definitions that utilize the `"oneOf": [{"required": ["<property name>"]}]` pattern, such as [`affects.items.properties.versions.items.oneOf`](https://github.com/CycloneDX/specification/blob/62a669075f1897193a14060e0784e6a7576b693d/schema/bom-1.6.schema.json#L2907).

The license definition says that only one of `id` or `name` is required, but both are permitted. Is this what was intended, or should these properties be mutually exclusive?

This validates against the schema, but does it match the intended usage of the property?

```json
"licenses": [
  {
    "license": {
      "id": "Apache-2.0",
      "name": "MIT"
    }
  }
]
```

Or might it make more sense to define the schema such that only one of `id` or `name` is allowed, for example by moving the `id` and `name` property definitions into their respective `oneOf` blocks?

<details>

<summary>Schema example</summary>

```jsonc
    "license": {
      "type": "object",
      "oneOf": [
        {
          "required": ["id"],
          "properties": {
            "id": {
              "$ref": "spdx.schema.json",
              "description": "A valid SPDX license ID"
            }
          }
        },
        {
          "required": ["name"],
          "properties": {
            "name": {
              "type": "string",
              "description": "If SPDX does not define the license used, this field may be used to provide the license name"
            }
          }
        }
      ],
      "additionalProperties": false,
      "properties": {
        "bom-ref": {
          "$ref": "#/definitions/refType",
          "description": "An optional identifier which can be used to reference the license elsewhere in the BOM. Every bom-ref MUST be unique within the BOM.\nValue SHOULD not start with the BOM-Link intro 'urn:cdx:' to avoid conflicts with BOM-Links."
        },
        // ...
      }
    }
```

</details>

**Comment by jkowalleck** [comment, marked as answer]
The intention is to have exactly one of `name` or `id` present. 
And that is exactly what the current JSON schema does. 


The JSON schema: https://github.com/CycloneDX/specification/blob/55343ba19dee1785acf1ce9191540d5fd7b590db/schema/bom-1.6.schema.json#L1232-L1258


For comparison, see XML schema: https://github.com/CycloneDX/specification/blob/55343ba19dee1785acf1ce9191540d5fd7b590db/schema/bom-1.6.xsd#L761-L772 

----

> This validates against the schema, 

Nope, it does not. maybe your json schema validator is implemented wrong?

JSON is valid against more than one schema from 'oneOf' - therefore, the data set you provided is invalid.



<details><summary>data used for testing the JSON schema</summary>
<p>

could be used on <https://www.jsonschemavalidator.net/>

schema cutout
```json
{
	"$schema": "http://json-schema.org/draft-07/schema#",
    "type": "object",
    "title": "License",
    "oneOf": [
        {
          "required": ["id"]
        },
        {
          "required": ["name"]
        }
    ],
    "additionalProperties": false,
    "properties": {
        "id": {  },
        "name": { },
    }
}
``` 

example data set 
```json
{

  "id": true,
  "name": true
}
```

</p>
</details>

**Comment by jkowalleck** [comment]
Additional technical detail:

> Or might it make more sense to define the schema such that only one of `id` or `name` is allowed, for example by moving the `id` and `name` property definitions into their respective `oneOf` blocks?

We found this to be not-working.
Having ` "additionalProperties": false, ` set on "root" level of a property would not allow additional properties that were not defined already -  Therefore, putting rules in an `oneOf` fails.

**Comment by jhoward-lm** [comment]
Ahh...my mistake. Thanks

**Reply by jkowalleck** [reply]
you are welcome. you are not the first with this specific question, so it is good to have it published here :+1:
